File: java/src/main/java/com/ngt/streamingwithflink/_05_CoProcessFunctionTimers.java
package com.ngt.streamingwithflink;

import com.ngt.streamingwithflink.util.SensorReading;
import com.ngt.streamingwithflink.util.SensorSource;
import org.apache.flink.api.common.state.ValueState;
import org.apache.flink.api.common.state.ValueStateDescriptor;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.co.CoProcessFunction;
import org.apache.flink.util.Collector;

import java.util.Arrays;

/**
 * @author ngt
 * @create 2021-05-18 9:05
 */
public class _05_CoProcessFunctionTimers {
	public static void main(String[] args) throws Exception {
		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
		env.getConfig().setAutoWatermarkInterval(1000L);

		DataStreamSource<Tuple2<String, Long>> filterSwitches = env.fromCollection(Arrays.asList(
				Tuple2.of("sensor_2", 5 * 1000L), Tuple2.of("sensor_7", 10 * 1000L)));

		DataStreamSource<SensorReading> readings = env.addSource(new SensorSource());

		readings.connect(filterSwitches)
				.keyBy(data -> data.id, data -> data.f0)
				.process(new ReadingFilter())
				.print();

		env.execute();
	}

	static class ReadingFilter extends CoProcessFunction<SensorReading, Tuple2<String, Long>, SensorReading> {

		private transient ValueState<Boolean> forwardingEnabled;
		private transient ValueState<Long> disableTimer;

		@Override
		public void open(Configuration parameters) throws Exception {
			forwardingEnabled = getRuntimeContext().getState(new ValueStateDescriptor<Boolean>("filterSwitch", Boolean.class));
			disableTimer = getRuntimeContext().getState(new ValueStateDescriptor<Long>("timer", Long.class));
		}

		@Override
		public void processElement1(SensorReading value,
									Context ctx,
									Collector<SensorReading> out) throws Exception {
			// Java版本中要注意为空的情况
			Boolean forward = forwardingEnabled.value();
			if (forward != null && forward) {
				out.collect(value);
			}
		}

		@Override
		public void processElement2(Tuple2<String, Long> value,
									Context ctx,
									Collector<SensorReading> out) throws Exception {
			forwardingEnabled.update(true);
			Long curTimerTimestamp = disableTimer.value();
			long timerTimestamp = ctx.timerService().currentProcessingTime() + value.f1;
			if (curTimerTimestamp == null || timerTimestamp > curTimerTimestamp) {
				if (curTimerTimestamp != null) {
					ctx.timerService().deleteProcessingTimeTimer(curTimerTimestamp);
				}
				ctx.timerService().registerProcessingTimeTimer(timerTimestamp);
				disableTimer.update(timerTimestamp);
			}
		}

		@Override
		public void onTimer(long timestamp, OnTimerContext ctx, Collector<SensorReading> out) throws Exception {
			forwardingEnabled.clear();
			disableTimer.clear();
		}
	}
}

File: scala/src/main/scala/com/ngt/streamingwithflink/_05_CoProcessFunctionTimers.scala
package com.ngt.streamingwithflink

import com.ngt.streamingwithflink.util.{SensorReading, SensorSource}
import org.apache.flink.api.common.state.{ValueState, ValueStateDescriptor}
import org.apache.flink.streaming.api.scala.{DataStream, StreamExecutionEnvironment}
import org.apache.flink.api.scala._
import org.apache.flink.api.scala.typeutils.Types
import org.apache.flink.streaming.api.functions.co.CoProcessFunction
import org.apache.flink.util.Collector

/**
 * @author ngt
 * @create 2021-05-11 10:09
 * 为不同的id设置不同的转发时间
 */
object _05_CoProcessFunctionTimers {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment

    env.getConfig.setAutoWatermarkInterval(1000L)
    val filterSwitches: DataStream[(String, Long)] = env
      .fromCollection(Seq(
        ("sensor_2", 5 * 1000L), // 转发 5 秒之后停止
        ("sensor_7", 10 * 1000L)) // 转发 10 秒之后停止
      )

    val readings: DataStream[SensorReading] = env.addSource(new SensorSource)

    readings
      .connect(filterSwitches)
      .keyBy(_.id, _._1)
      .process(new ReadingFilter)
      .print()

    env.execute()

  }
}

class ReadingFilter extends CoProcessFunction[SensorReading, (String, Long), SensorReading] {

  private lazy val forwardingEnabled: ValueState[Boolean] =
    getRuntimeContext.getState(new ValueStateDescriptor[Boolean]("filterSwitch", classOf[Boolean]))


  private lazy val disableTimer: ValueState[Long] =
    getRuntimeContext.getState(new ValueStateDescriptor[Long]("timer", classOf[Long]))

  override def processElement1(value: SensorReading,
                               ctx: CoProcessFunction[SensorReading, (String, Long), SensorReading]#Context,
                               out: Collector[SensorReading]): Unit = {
    if (forwardingEnabled.value()) {
      out.collect(value)
    }
  }

  override def processElement2(value: (String, Long),
                               ctx: CoProcessFunction[SensorReading, (String, Long), SensorReading]#Context,
                               out: Collector[SensorReading]): Unit = {

    forwardingEnabled.update(true) // 开启读数转发

    val curTimerTimestamp: Long = disableTimer.value()
    val timerTimestamp: Long = ctx.timerService().currentProcessingTime() + value._2

    if (timerTimestamp > curTimerTimestamp) {
      // 移除当前计时器创建一个新的计时器
      ctx.timerService().deleteProcessingTimeTimer(curTimerTimestamp)
      ctx.timerService().registerProcessingTimeTimer(timerTimestamp)
      disableTimer.update(timerTimestamp)
    }
  }

  override def onTimer(timestamp: Long,
                       ctx: CoProcessFunction[SensorReading, (String, Long), SensorReading]#OnTimerContext,
                       out: Collector[SensorReading]): Unit = {
    // 移除所有状态，默认情况下转发开关关闭
    forwardingEnabled.clear()
    disableTimer.clear()
  }
}